### ISSUE TITLE
WebRTC support

### ISSUE THREAD

**Issue by ken0x0a** [tauri-apps/wry#85]
I tested only on macOS.
It might already work on other platforms.

**Is your feature request related to a problem? Please describe.**
I want to use WebRTC

**Describe the solution you'd like**
I don't have any preference.
I just want to use WebRTC.

**Describe alternatives you've considered**
- With system webbrowser (doesn't work on macOS)
  ```js
  Notification.requestPermission() // => undefined
  navigator.mediaDevices // => undefined
  ```
- Using another process
  I don't know any library to use WebRTC with VideoUI on desktop 😭

**Additional context**

It seems, the cause of this issue is this https://forums.developer.apple.com/thread/88052

And, the following might solve this issue on macOS (not tested).
https://bugs.webkit.org/show_bug.cgi?id=208667#c17

---

There are similar issues in other repositories.

- https://github.com/patr0nus/DeskGap/issues/26#issuecomment-486463510
  (someone mentioned "WkWebView's getUserMedia functionality is disabled, when embedded")
- https://github.com/zserge/webview/issues/368

**Comment by wusyong** [comment]
Maybe this is also belongs to #81?

**Comment by nklayman** [comment]
WebRTC also is not supported by `Webkit2GTK`. It technically doesn't require permissions to work, but it's most commonly used with a webcam/mic input and those require permissions.

**Comment by lemarier** [comment]
### macOS update

The runtime requires some permissions. Adding an embed PLIST to the runtime fixed the issue for me.

You can give it a try with [embed_plist](https://crates.io/crates/embed_plist)

Sample PLIST
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>NSCameraUsageDescription</key>
	<string>Request camera access for WebRTC</string>
	<key>NSMicrophoneUsageDescription</key>
	<string>Request microphone access for WebRTC</string>
</dict>
</plist>
```

Test app
```rust
#[cfg(target_os = "macos")]
embed_plist::embed_info_plist!("Info.plist");

fn main() -> wry::Result<()> {
  use wry::{
    application::{
      event::{Event, StartCause, WindowEvent},
      event_loop::{ControlFlow, EventLoop},
      window::WindowBuilder,
    },
    webview::WebViewBuilder,
  };

  let event_loop = EventLoop::new();
  let window = WindowBuilder::new()
    .with_title("Hello World")
    .build(&event_loop)
    .unwrap();
  let _webview = WebViewBuilder::new(window)
    .unwrap()
    .with_url("https://test.webrtc.org/")?
    .build()?;

  event_loop.run(move |event, _, control_flow| {
    *control_flow = ControlFlow::Poll;

    match event {
      Event::NewEvents(StartCause::Init) => println!("Wry application started!"),
      Event::WindowEvent {
        event: WindowEvent::CloseRequested,
        ..
      } => *control_flow = ControlFlow::Exit,
      _ => (),
    }
  });
}
```

Also, <s>seems to works only over HTTPS, so doesn't work for custom protocol</s>, so I think the wry job is pretty much done, <s>only require the custom protocol support who I think should be moved to another issue</s>.

Tauri should give the possibility to inject a PLIST at runtime for development. For bundled app, we already generate a plist, so maybe we should add it as a configuration object to request `microphone`, `webcam` access, so we could inject it automatically.

Edit: Looks like I've been able to get custom protocol to work. We need to make sure to use the host `localhost` in our custom protocol. So for example, instead of using `wry://` as the entry point, we should use `wry://localhost/`, this way content, is processed as HTTPS.

**Comment by wusyong** [comment]
It seems webkitgtk doesn't support webrtc.
https://bugs.webkit.org/show_bug.cgi?id=194805

**Comment by wusyong** [comment]
Found a bug tracker states webkitgtk does has webrtc support.
https://bugs.webkit.org/show_bug.cgi?id=196222

I'll try to build it this night.

**Comment by nothingismagick** [comment]
Regarding boring-ssl and its gpl nature, we aren't shipping it (just linking) so it will not impact us.

**Comment by wusyong** [comment]
yeah I think you are right. We cannot distribute it for non-GPL apps because of it.
Below is the status report from WPE but should hold the same to WebkitGtk
https://wpewebkit.org/about/faq.html#what%E2%80%99s-the-status-regarding-webrtc%3F

**Comment by VarLad** [comment]
https://webrtc.rs/ seems usable now
Can that be somehow used for Tauri?

**Comment by YuilTripathee** [comment]
Is there any solution connecting other languages implementation like SWIG?
Seems like a workaround (but for temporary use might be usable)

**Comment by fxtressia** [comment]
Hiiii ^^

Can we rename this issue to "WebRTC support on Unices"?

It seems this issue is also talking about MacOS (which isn't Linux of course), soo it'll be great for clarity!!!

**Comment by FabianLars** [comment]
@lovelyyfiaaa Heyy :)
webrtc does work on macos, that's why the issue was renamed from `WebRTC support` to `WebRTC support on Linux` a few messages above (actually it was fixed a little bit before it was renamed)

**Comment by fxtressia** [comment]
> @lovelyyfiaaa Heyy :) webrtc does work on macos, that's why the issue was renamed from `WebRTC support` to `WebRTC support on Linux` a few messages above (actually it was fixed a little bit before it was renamed)

Aaah, I seee. But I think we should do something to make this issue less confusing?

**Comment by fxtressia** [comment]
> https://webrtc.rs/ seems usable now Can that be somehow used for Tauri?

Umm, how does it support Pipewire? It's required in Linux because of security requirements implemented in Wayland and Flatpak ^^

**Comment by lovelyyfiaaa** [comment]
Ummm,  okayy!! There's no mention for it in github.dev...

![Screenshot from 2022-08-03 19-44-45](https://user-images.githubusercontent.com/66000635/182611016-7ed8b3fd-c5bf-4740-bd09-48885794e670.png)

**Comment by fxtressia** [comment]
![image](https://user-images.githubusercontent.com/66000635/182611340-f754efe2-983e-4ba0-89bd-9cb3c69162d8.png)

Ahh okay!! The codebase has no mention of it, so we should ask the devs first ^^

**Comment by fxtressia** [comment]
I'm opening https://github.com/webrtc-rs/webrtc/issues/231 nowww <3

**Comment by FabianLars** [comment]
I think a relevant issue for this would be: <https://github.com/webrtc-rs/webrtc/issues/94> and the crates linked in that, none of them seem to _explicitly_ support pipewire.

Also, unless i got it completely wrong, this issue shows that it's not ready for tauri at all, which means we have to keep working for webkitgtk, buuut last time i checked, they planned on enabling the gstreamer backend in the 2.38 :)

**Comment by fxtressia** [comment]
> Regarding boring-ssl and its gpl nature, we aren't shipping it (just linking) so it will not impact us.

Ummm, I'm not a lawyer, buuut doesn't linking too trigger the GPL? 

For proprieritary softwares, this is safe, as long as the company making the proprieritary part isn't hostile (for example Minecraft modders distribute mod patches and not the modified binaries), but this isn't the same case for crowd-enforced open source licenses right?

**Comment by fxtressia** [comment]
> Also, unless i got it completely wrong, this issue shows that it's not ready for tauri at all, which means we have to keep working for webkitgtk, 

Yep!! There's https://github.com/RustAudio/cpal/issues/554 but it seems it's not a finished issue yet?

**Comment by fxtressia** [comment]
Umm, what do you guys think about using QtWebEngine?

**Comment by FabianLars** [comment]
> Umm, what do you guys think about using QtWebEngine?

<https://github.com/tauri-apps/tauri/issues/4554>

**Comment by fxtressia** [comment]
Aha, yup!!! There isn't a good bindings library for Qt in Rust yet for noww. ^^

**Comment by yuluo-zy** [comment]
Is WebRTC support available on Linux ?

**Comment by FabianLars** [comment]
@yuluo-zy No, but we're somewhat optimistic that the next webkit2gtk version will add support for webrtc.

**Comment by FabianLars** [comment]
Oh i forgot to update my above message - Yes, initial support was added in 2.38 but it's hidden behind a build flag and so far not a single distro enables it. Therefore we unfortunately have to keep waiting...

**Comment by zippy** [comment]
Is there any action on this issue?  I have just run into the same thing, and would really like this to work in Tauri.

**Comment by FabianLars** [comment]
My last comment is still accurate. We're still waiting for the next webkigtk update which hopefully has more complete webrtc support. (they have a ~6 month release schedule to always get a new release out before a new ubuntu version in case you're wondering)

**Comment by DougAnderson444** [comment]
> I am wondering if we will have to roll our own: [libp2p/rust-libp2p#1066](https://github.com/libp2p/rust-libp2p/issues/1066)

WebRTC was merged into libp2p (for server-to-browser for now, which should "serve" Tauri's needs 😆) with this PR:

https://github.com/libp2p/rust-libp2p/pull/2622 (Merged 🎉)

**Comment by luke-goddard** [comment]
Webkitgtk release note is a bit cryptic, can someone more informed than me confirm if WebRTC is supported with this release? 
[WebKitGTK 2.38.4 released!](https://webkitgtk.org/2023/02/02/webkitgtk2.38.4-released.html) - Fix the build with GStreamer-based WebRTC enabled.

**Comment by FabianLars** [comment]
Nope, it's hidden behind a (experimental?) build-time flag that no distro enabled so far, which means we have to wait for it to be stable and enabled by default on the webkitgtk side.

**Comment by edualb** [comment]
@FabianLars do you have any update about this topic? Is there anything we (as community) can help with it?

**Comment by FabianLars** [comment]
> do you have any update about this topic?

Nothing specific, afaik it's making steady progress but i didn't see any hints about it being enabled by default in 2.42 yet (or any kind of ETAs for that matter).

> Is there anything we (as community) can help with it?

People with experience in low-level stuff that may know a thing or two about anything that may help the webkitgtk project (whatever that would be) could reach out to the webkitgtk devs and see if they need any help.
Apart from that we can only really be patient.

**Comment by learncodingforweb** [comment]
unable to find **navigator.mediaDevices**. Any update?

**Comment by thenbe** [comment]
Instructions for how to enable webRTC on nixOS can be found in this [thread](https://github.com/tauri-apps/tauri/discussions/8426#discussioncomment-8268622). It's still a bit rough around the edges, but it works.

**Comment by JIGutierrez** [comment]
Have there been any workarounds discovered for this, specifically around P2P? I've been trying to use `simple-peer` on a Linux build, and haven't had any success with injecting webRTC packages, and I'd like to avoid changing protocols if possible...

Fingers crossed 🤞🏽

**Comment by DougAnderson444** [comment]
@JIGutierrez for my p2p tauri apps I [use the WebRTC transport](https://github.com/libp2p/rust-libp2p/tree/master/transports/webrtc) built into `rust-libp2p`

**Comment by JIGutierrez** [comment]
> @JIGutierrez for my p2p tauri apps I [use the WebRTC transport](https://github.com/libp2p/rust-libp2p/tree/master/transports/webrtc) built into `rust-libp2p`

Right, but as far as I'm aware (new to Tauri and Rust) there's no way of using that JS-Side unless you create bindings for it, right?

**Comment by DougAnderson444** [comment]
What do you want to do exactly? Control your Rust WebRTC node from the JavaScript front end of Tauri? Relay Rust-Tauri WebRTC messages to the JS front end? Communicate with other WebRTC nodes (other people) from the JS front end of Tauri?

**Comment by JIGutierrez** [comment]
> What do you want to do exactly? Control your Rust WebRTC node from the JavaScript front end of Tauri? Relay Rust-Tauri WebRTC messages to the JS front end? Communicate with other WebRTC nodes (other people) from the JS front end of Tauri?

I'd like to use the js library `simple-peer`, which uses the browser's WebRTC API, which is not in webgtk2 on Linux. You can optionally supply a custom webRTC implementation, but the only ones available are for Node and don't work on Tauri (as far as I've tried). I was considering compiling webrtc-rs to webassembly (haven't tried to know if it's compatible), but that does seem like overkill.

**Comment by DougAnderson444** [comment]
I think you may be mixing different contexts here. You have 2 parts to webRTC: The browser client(s) and the server.

You can indeed use `simple-peer` in the JavaScript front-end of your Tauri app for the browser client. That's no problem. You're right in that `simple-peer` can't be used as the server, and `webrtc-rs` cannot be compiled to wasm for the browser.

To connect two browser clients, traditionally you will need a signalling server to use `simple-peer`, which you can technically use your Tauri backend for as your server, which is what I assume you're going for. You'd likely set it up as a Nodejs sidecar, or [find a Rust crate](https://crates.io) to call from main. However, keep in mind that signalling servers are typically websocket based, and for secure websockets (wss://) you'll need TLS for which you'll a domain name and Let's Encrypt. Traditional signalling servers likely aren't a great fit for home Tauri apps for this reason but you could probably do it.

Instead of connecting to a wss signalling server, you could set up Tauri app as a WebRTC server, which is what `webrtc-rs` would give you. This has [the advantage](https://blog.libp2p.io/rust-libp2p-browser-to-server/#easier-node-operation) that we **don't need TLS, domain name, or Let's Encypt** as the certificate is self-signed. You would normally connect the client to the server, but I don't believe `simple-peer` is configured to connect to `webrtc-rs` as it's [not typically used](https://github.com/webrtc-rs/webrtc/blob/master/examples/examples/README.md) a as a signalling server.

Is there a reason you're set on `simple-peer`?

[I have a Tauri app](https://github.com/PeerPiper/peerpiper-desktop?tab=readme-ov-file#peerpiper-desktop) running [`rust-libp2p`](https://github.com/libp2p/rust-libp2p) as the WebRTC server node, to which browser peers can connect to over WebRTC. It's `libp2p` based (instead of `simple-peer`), and does only data (no media/video), but sounds more like what you're intending on using Tauri for? You want to connect browsers to your Tauri app, yes?

**Comment by matthme** [comment]
> You can indeed use simple-peer in the JavaScript front-end of your Tauri app for the browser client. That's no problem.

I think this is in fact *exactly* the problem this whole github issue is about, that this is not possible on Linux. I am no world-class systems builder however, so I might be dissecting the problem space along the wrong axes...

**Comment by bra1n** [comment]
The webkit devs compiled an entry in their FAQ for [when WebRTC support arrives in Webkit](https://wpewebkit.org/about/faq.html#what%E2%80%99s-the-status-regarding-webrtc%3F) and it links to [this issue](https://bugs.webkit.org/show_bug.cgi?id=235885). So it seems like there is still some way to go before full WebRTC support arrives for Webkit and with that for Tauri on Linux.

**Comment by nyabinary** [comment]
So what the status with this and Tauri v2

**Comment by FabianLars** [comment]
Still the same. Also, it's not coupled to the Tauri version (if we ignore https://github.com/tauri-apps/tauri/issues/9662).

**Comment by j4gd33p** [comment]
Any update on this ? is this supported on chromeOs (Linux) ?

**Comment by FabianLars** [comment]
> Any update on this ?

No

> is this supported on chromeOs (Linux) ?

i don't think tauri supports chromeos in general, does it?

**Comment by j4gd33p** [comment]
Yes, my app is working on chrome OS except webrtc part(webrtc - 60% works). 
BUT there is a point Chrome crostini runs in a linux container, Its like a VM. So tauri have access to VM functions  only, not chrome OS.
After doing research & investigations we shifted to chrome OS IWA (isolated web app) and able to share screens and do our work.

**Comment by tushar1977** [comment]
is webRTC supported in linux?

**Comment by FabianLars** [comment]
no (wondering if i should have a auto-reply bot here that just says `no` after each comment 😂 )

**Comment by bra1n** [comment]
> no (wondering if i should have a auto-reply bot here that just says `no` after each comment 😂 )

Has there been any progress made on this? Or are y'all basically waiting for https://bugs.webkit.org/show_bug.cgi?id=235885 to be finished first?

**Comment by FeldrinH** [comment]
There are other avenues to fix this, such as https://github.com/tauri-apps/tauri/issues/14963.

**Comment by tushar1977** [comment]
> no (wondering if i should have a auto-reply bot here that just says `no` after each comment 😂 )
😂 bruhh 😔

**Comment by chxreyes** [comment]
Is there anyone actively working on implementing this?

**Comment by FeldrinH** [comment]
That kinda depends on what you mean by "this". I don't know if anyone is actively working on improving WebRTC support in WebKit, but there is active work on integrating CEF into Tauri (see https://github.com/tauri-apps/tauri/issues/14963#issuecomment-3915824782), and CEF has good WebRTC support on Linux, macOS and any other platform that it supports.

**Comment by Techognito** [comment]
> > no (wondering if i should have a auto-reply bot here that just says `no` after each comment 😂 )
> 
> Has there been any progress made on this? Or are y'all basically waiting for https://bugs.webkit.org/show_bug.cgi?id=235885 to be finished first?

That bug was closed with wontfix tag, and https://bugs.webkit.org/show_bug.cgi?id=244004 was merged instead (from what I understand)

> is webRTC supported in linux?

With Tauri/webkit no, with Electron/CEF yes. (With Tauri/CEF probably)



> There are other avenues to fix this, such as [tauri-apps/tauri#14963](https://github.com/tauri-apps/tauri/issues/14963).

This feels more like a workaround than a fix, as I personally think of tauri as an alternative to electron and it's use of CEF. That said I understand the choice of changing to CEF, I just wish it wasn't needed.
